Lib/test/test_types.py

```diff
@@ -703,22 +703,32 @@ def test_hash(self):
         self.assertEqual(hash(int | str), hash(str | int))
         self.assertEqual(hash(int | str), hash(typing.Union[int, str]))
 
-    def test_instancecheck(self):
-        x = int | str
-        self.assertIsInstance(1, x)
-        self.assertIsInstance(True, x)
-        self.assertIsInstance('a', x)
-        self.assertNotIsInstance(None, x)
-        self.assertTrue(issubclass(int, x))
-        self.assertTrue(issubclass(bool, x))
-        self.assertTrue(issubclass(str, x))
-        self.assertFalse(issubclass(type(None), x))
-        x = int | None
-        self.assertIsInstance(None, x)
-        self.assertTrue(issubclass(type(None), x))
-        x = int | collections.abc.Mapping
-        self.assertIsInstance({}, x)
-        self.assertTrue(issubclass(dict, x))
+    def test_instancecheck_and_subclasscheck(self):
+        for x in [int | str, typing.Union[int, str]]:
```

Comment:
Interesting, is this supposed to work? I know PEP 604 specified that isinstance() should work for int | str, but where does the implementation for `typing.Union` come from?

Reply:
It was added here: https://github.com/python/cpython/blame/08f8301b21648d58d053e1a513db8ed32fbf37dd/Lib/typing.py#L1243-L1249

Reply:
I think it is a good idea that it works this way. Why?
1. It would be strange for `int | str` to work and for `Union[int, str]` not to work
2. The semantics is pretty clear. I would expect it to work if I was a new-commer

Reply:
Maybe, but we made a general decision a long time ago not to support isinstance() on typing.py primitives. `isinstance(..., Any)` doesn't work for example.

That said, this behavior was released in 3.10 so I guess we're stuck with it now, so we might as well test it.

Could you add a test to make sure it errors correctly when the Union includes something that can't be isinstance()d, such as a TypeVar?

Reply:
@JelleZijlstra looks like you found a new bug! Consider this:

```python
>>> from typing import Union, TypeVar
>>> T = TypeVar('T')
>>> issubclass(int, Union[int, T])
True
>>> issubclass(int, Union[T, int])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sobolev/Desktop/cpython/Lib/typing.py", line 1248, in __subclasscheck__
    if issubclass(cls, arg):
       ^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 2 must be a class, a tuple of classes, or a union
```

Reply:
One way to solve this is to always check **all** `Union` items with `issubclass()`. Slower, but more reliable.

Reply:
I don't think that's a bug, it's the same with `|`-based unions (see test_bad_instancecheck below). But we should test this behavior.